CORE: Added Gen2 Multi Domain Vehicle Taxonomy

danielwilms opened the pull request:
* Added and described the term
* Restructured the chapters
* Added skeleton for further description

[Readable Version](https://raw.githack.com/w3c/automotive/9d4c68de97195eec868fbd5875cc5821a32e54f8/spec/Gen2_Core.html)

Fixes: #306

danielwilms committed:
CORE: Added Gen2 Multi Domain Vehicle Taxonomy

* Added and described the term
* Restructured the chapters
* Added skeleton for further description

Fixes: #306

PatrickCQ commented on spec/Gen2_Core.html:
```diff
@@ -75,130 +80,333 @@ <h2>Introduction</h2>
+                    Please note that, from a YAML perspective, the include directive is just 
+                another comment.
+            </section> 
+            
```
Describes a schema file.

PatrickCQ commented on spec/Gen2_Core.html:
```diff
@@ -75,130 +80,333 @@ <h2>Introduction</h2>
+        
+        <section id="datadefinition">
+            <h3>Data Definition</h3>
+            The Data Definition is the representation of the specific domain under using the Rule Set described before.
```
Definition of *a* specific domain (the vehicle data/body/driving domain).

PatrickCQ commented on spec/Gen2_Core.html:
```diff
@@ -75,130 +80,333 @@ <h2>Introduction</h2>
-          <p>A leaf the same as <a>sensor</a> which additionaly can be changed.</p>
+        
+        <section id="datadefinition">
+            <h3>Data Definition</h3>
```
An example for a Domain.

danielwilms replied:
You mean I should change the title? I would hope that we add domains here before we finish the spec, that's why I defined it like this. What do you think?

PatrickCQ replied:
> You mean I should change the title? I would hope that we add domains here before we finish the spec, that's why I defined it like this. What do you think?

Adding multiple domains here sounds good, but it felt to me to be build for only one. If you could at something that makes that more clear that would be very much appreciated.

danielwilms replied:
> Adding multiple domains here sounds good, but it felt to me to be build for only one. If you could at something that makes that more clear that would be very much appreciated.

Let’s have a goal to get the multimedia domain there. I checked your mail, but I think it's hard to discuss via mail. Would you agree that I add your vspec file to my clone of this repo? Then we can take it from there.

danielwilms replied:
@PatrickCQ: In the last commit I added some more description for the data definition.

AaronRobson replied:
The minimum value for a signed 64-bit integer is `-(2^63)`, you have the positive.

PatrickCQ reviewed (changes requested):
Suggestion, change to "schema file" instead of Basic Rules.
Make clear that Data Definition is an example for the Domain Vehicle Signals.

UlfBj reviewed (approved):
In figure 1 the name of the root node should be Vehicle.
The datatypes Float and Double should maybe have references to their exact definitions, as these are not uniquely defined by their names only?

danielwilms commented:
> In figure 1 the name of the root node should be Vehicle.

Good point @UlfBj. Changed it.

> The datatypes Float and Double should maybe have references to their exact definitions, as these are not uniquely defined by their names only?

Where do you mean exactly?

danielwilms committed:
CORE: Fixed comments on PR #315

UlfBj commented:
>> The datatypes Float and Double should maybe have references to their exact definitions, as these are not uniquely defined by their names only?
Where do you mean exactly?

I had forgot the details of the float/double format so I looked it up, and it said most computers followed this standard: https://en.wikipedia.org/wiki/IEEE_754
So maybe it would be good to refer to it? If so, where these are mentioned in the chapter. 
It is fine with me to skip it, if "most computers" means almost all computers.

wzr1337 commented:
In order to make it exact how about float32 float64 and float128 to express
float, double and decimal respectively?

On Thu, 7 Nov 2019 at 19:47 Ulf Björkengren <notifications@github.com>
wrote:

> The datatypes Float and Double should maybe have references to their exact
> definitions, as these are not uniquely defined by their names only?
> Where do you mean exactly?
>
> I had forgot the details of the float/double format so I looked it up, and
> it said most computers followed this standard:
> https://en.wikipedia.org/wiki/IEEE_754
> So maybe it would be good to refer to it? If so, where these are mentioned
> in the chapter.
> It is fine with me to skip it, if "most computers" means almost all
> computers.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/w3c/automotive/pull/315?email_source=notifications&email_token=AA4IL3NY3G4EVSAGCXJPAWTQSRPDXA5CNFSM4JKGIFOKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDNNHCQ#issuecomment-551211914>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4IL3IN2OLJAGXCPMJZEB3QSRPDXANCNFSM4JKGIFOA>
> .
>
-- 
- Patrick (Mobile)

PatrickCQ commented:
> In order to make it exact how about float32 float64 and float128 to express float, double and decimal respectively?

the data types are not really part of the changes @danielwilms proposed. He just copied that part around to a better location in the document. In my opinion we should discuss data types in a separate issue to not overload this proposed change.

danielwilms committed:
CORE: Explixitly mentioning the possibility of multiple domains

Extending PR #315

PatrickCQ commented on spec/Gen2_Core.html:
```diff
@@ -307,8 +307,12 @@ <h3><dfn>ByteArray</dfn></h3>
+                The Data Definition is the representation of a specific domain under using the <a href="#ruleset">Rule Set</a>.
+            </p>
+            <p>
+                The Vehicle Domain Taxonomy SHALL be the base, however it MAY be extended by taxonomies outside the vehicle signals domain. Added taxonomies MUST follow the <a href="#ruleset">defined</a> rules.
```
I do not agree with the base. That would request every implementer to at least have the vehicle domain taxonomy. 
"The vehicle domain taxonomy is an example. 
Other taxonomies outside of the vehicle signals domain, e.g. multimedia or navigation, are possible as well.
Added taxonomies MUST follow the <a href="#ruleset">defined</a> rules."

danielwilms replied:
>  do not agree with the base. 

That was the conclusion of #306. I think the question is, how to align those. 

> The vehicle domain taxonomy is an example.

This wording would be too weak for me. I think we have to bring it up in the call. 
Or @gunnarx, @UlfBj, @tguild and others. Any opinion on this one?

UlfBj replied:
My understanding of what we have agreed on in the WG is that VSS is the data model that Gen2 will recommend and use in examples, and that Gen2 shall have a Vehicle Domain focus. 
With that said, I think the current wording is good. 
To further clarify, I think that the VSS spec shall require the Vehicle Domain as base. Then in Gen2 we *recommend* usage of VSS (and use it for examples). 
So if Gen2 shall include a copy of parts of the VSS spec, then that should be identical to the "real" VSS spec, or else it would be very confusing. 
So my view is that the VSS (*Vehicle Signal* Specification) shall be clear on having the Vehicle Domain as base, while Gen2 recommends usage of VSS. 
As a side note I think this provides an argument for not including a copy of the VSS spec in Gen2, but rather referring to it, and possibly provide a high level overview.

peterMelco replied:
In respect to that, what does then statement "The Vehicle Domain Taxonomy SHALL be the base" really mean for an implementer. I think Patrick has a point in that this statement locks an implementer into at least The Vehicle Domain Taxonomy, or so it seems. So, in my view then the wording should reflect that VSS "data definition" is an example. I don't think the wording makes that clear enough.

UlfBj replied:
As I said above, with Gen2 recommending usage of VSS, then VSS can state "The Vehicle Domain Taxonomy SHALL be the base" without that leading to an implementer of Gen2 to be locked into the VSS Vehicle Domain Taxonomy, but instead decide to use some other data model. 
Maybe I am wrong in seeing VSS, VISS and Gen2 as Vehicle Domain focused specifications, but if I am not, then watering down VSS to not requiring the Vehicle Domain Taxonomy as base is the wrong way to go.

tguild replied:
Gen2 is meant to support multiple domains. For Vehicle signals, yes the group agrees on VSS and ensure Gen2 can support it. We need to strike a balance emphasizing VSS but make clear it is not exclusively about VSS. For instance I can see a Gen2 in-vehicle service instance supporting data (traffic, weather, road conditions, notifications) that originated outside the vehicle and service providing a maintained cache.

UlfBj replied:
@tguild , I think your comment is fully supported by my proposal in the previous comment. VSS has the ability to support other domains like what you exemplify. For any Gen2 adopter that do not like VSS there is the possibility to replace VSS with another data model. 

>> We need to strike a balance emphasizing VSS but make clear it is not exclusively about VSS. 

This is the key point, to find how to express this in the data model chapter. 
Personally I think this "balance" is easier to strike if we do not copy large pieces of the VSS description into this spec, but rather refer to it, with only a high level overview. By going into detail in its description, other alternative data models would need similar detail to achieve "balance". 
Describing the requirements a data model must fulfill to be able to be used in Gen2, recommending VSS, and providing references to possible alternatives would in my view strike a good balance.

tguild replied:
I do not think VSS (*Vehicle Signal* Specification) should represent other domains besides vehicle signals. The taxonomy and rules should be clear so what is expected and supportable is understood for those bringing additional data models they want to make available in Gen2.

peterMelco replied:
Could we not just add ...** For vehicle signals** the vehicle domain taxonomy SHALL be the base. The taxonomy can be either extended or replaced by other domains outside the vehicle signal domain. Added taxonomies must follow the <> rules. ?

PatrickCQ replied:
> ** For vehicle signals** the vehicle domain taxonomy SHALL be the base

Even mixing gen2 core and service is too controversial in my opinion. 
Our charter up to 2020 describes building (1) a common pattern good for uses within the vehicle **and** (2) typical vehicle services.
https://rawcdn.githack.com/w3c/automotive/89681830a855cfe9a896d112664048d28205cf7b/charter-2018.html#deliverables

tguild replied:
Two topics I want to discuss in this one phrase and a suggested way to phrase.

@UlfBj @danielwilms 
I find the notion of a single tree (branch per domain under a VSS root) as discussed on call and in this thread problematic for a number of reasons. What happens for instance if two underlying services claim the same branch? That is plausible with aftermarket use cases and an on-board app would likely want to be able to access both branches simultaneously. Separate Gen2 service per domain makes sense and believe we previously concluded this during VISS. The services can be well cataloged as part of a registry service so as to facilitate discovery for client apps, balance load, etc. An aftermarket device could register itself as a Gen2 service and provide new sets of signals readily or send those within the vehicle network (eg CAN) and appear as a VSS private branch in the OEM provided vehicle signal service. I can go into additional reasons on call or by email why a single root may not be the preferred solution.

Aftermarket (permissible under Right to Repair)  devices will want to be able to expose signals within the vehicle as well as offboard. More capable TPMS with additional sensors (eg basic microphones to measure vibrations) introduced later by a tire manufacturer combined with existing VSS enables life and cost saving algorithmic edge computing opportunities. For some signals they should clearly be available in the main Gen2 exposing VSS and either replacing existing signals directly or expose them on a private branch. Do we really want to have to manage that complexity under a single tree for all potential services on a vehicle?  Seems prone to conflicts.

Thinking of wording along the lines of:

This specification was created with VSS as the principal taxonomy for vehicle signals. In order to expose data through Gen2 it is expected taxonomies will adhere to a fundamental set of rules. [enumerate rules]

UlfBj replied:
>> What happens for instance if two underlying services claim the same branch?

In the dynamic registry proposal I presented on the WG, the proposal says that for the tree being "imported" from VSS and installed in the vehicle at manufacturing, there is in that install context one underlying service that registers for the entirety of this tree. Any following underlying services can only register for additions to this tree, they cannot claim any parts of the existing tree. 

>> Separate Gen2 service per domain makes sense and believe we previously concluded this during VISS. The services can be well cataloged as part of a registry service so as to facilitate discovery for client apps, balance load, etc.

I have added to my slide presentation 
(https://docs.google.com/presentation/d/1Eza-LNe8w7lJUA3YkhOUdoeDRkcrZo1645qsd8ehToc/edit?pli=1#slide=id.g6b21a63a7d_0_7)
a solution to this, please see slide 7. With this VSS can take the roll of a registry also for separate trees. 
With this the single root is a single entry point to the registry, which I think is desirable. 
The separate tree do not have to conform to the VSS Rule Set, only to what is required for it to be compatible with the Gen2 API. 

>> Do we really want to have to manage that complexity under a single tree for all potential services on a vehicle?

With the solution on slide 7 both alternatives are available for aftermarket use cases, either to attach it into the VSS tree, or keep it separate, and only include a pointer to the separate tree in VSS. So the aftermarket service provider can select what he thinks is the best alternative.

danielwilms replied:
Weakened the requirement, like discussed in the call on the 2019-11-19. Resolve the conversation with the commit.

danielwilms committed:
VSS as base for multidomain taxonomy

Weaken the requirement from SHALL to SHOULD following RFC2119:

"SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course."

danielwilms committed:
CORE: Changed SHALL to SHOULD

UlfBj commented:
After discussion with Daniel that the content of this PR has since its creation become obsolete in view of other accepted PRs, it is now closed.
